### DISCUSSION TITLE
Inserting with OneToOne association

### DISCUSSION THREAD

**Discussion by runningjustin** [babyfish-ct/jimmer#660] [Q&A]
Is it possible to create 2 objects in a OneToOne relationship via a single save command? I am not using a join table. For instance, when I create a User which is 1-1 with Person can I also create the person? This doesn't seem to work but perhaps my syntax is wrong.

I've tried the following:
```
sqlClient.entities.save(new(User::class).by {
    id = "12345"
    status = "ACTIVE"
    person = new(Person::class).by {
        id = "12345"
        title = "Mr."
        firstName = "John"
        lastName = "Doe"
        email = "john@doe.com"
        dateOfBirth = "2000-01-01"
    }})
```
I also tried:
```
sqlClient.entities.save(new(User::class).by {
    id = "12345"
    status = "ACTIVE"
    person  {
        id = "12345"
        title = "Mr."
        firstName = "John"
        lastName = "Doe"
        email = "john@doe.com"
        dateOfBirth = "2000-01-01"
    }})
```

`user_id` is nullable in the Person table and when I try to execute either of the above options I get this error:
`java.lang.IllegalArgumentException: The value "USER" cannot be parsed as JSON object"`
This is on version 0.8.132.

Any guidance would be greatly appreciated!

**Comment by runningjustin** [comment]
The error was coming from my test setup. Will re-raise the question if I run into other issues. Thanks!

**Comment by babyfish-ct** [comment]
Those two ways are equal and valid.

Please give your entity definition and full stack trace.

> BTW, the `0.8.170` is ready,save command is re-implemented(Huge refactor), by batching sql and database level upsert *(However, it must cost a few weeks to upgrade documentation, after that, a new version called `0.9.0` will be release on github)*

**Reply by runningjustin** [reply]
I got it sorted out, thanks! I have another question about the new save changes that I'll raise separately.

**Comment by runningjustin** [comment]
Reopening this as while my tests work because I'm using mock ids, at runtime I keep getting the error:

`Unexpected error: Save error caused by the path: "<root>.person": Cannot save illegal entity object whose type is "com.package.Person", entity with neither id nor key cannot be accepted. There are 3 ways to fix this problem: 1. Specify the id property "id" for save objects; 2. Specify the value key properties "user" for saved objects; 3. Specify the aggregate-root save mode of the save command to "INSERT_ONLY"(function changed) " or "NON_IDEMPOTENT_UPSERT".`

In my tests I'm able to pass userId in the person structure, but at runtime I won't know that id because it's generated dynamically upon insert (this rules out option 1). Also, of course, without having already saved the user I can't use option 2. I am using insert and have also tried save with INSERT_ONLY and NON_IDEMPOTENT_UPSERT. How can I update my syntax to make this work?

**Comment by babyfish-ct** [comment]
Neither Id Nor Key

- Id means **surrogate id**, one properties decorated by @Id
- Key means **natural id**, many properties(scalar or many-to-one) decorated by @key
 
Id is required, key is optional. When the id is auto generated so that the id can be missed, please configure @Key for entity type and specify the keys of objects

Neither id nor key means the idempotence cannot be guaranteed, so the new version throws an error *(Totally refactor, delete and reimplement)*.

However, the developer can 
```
save(objWithNeightIdNorKey) {
    setMode(SaveMode. NON_IDEMPOTENT_UPSERT)
}
```
to get the same default behavior of old version*(Old version is designed wrong)*

Please view the java doc of [aggregate-root save mode](https://github.com/babyfish-ct/jimmer/blob/main/project/jimmer-sql/src/main/java/org/babyfish/jimmer/sql/ast/mutation/SaveMode.java)

**Reply by babyfish-ct** [reply]
Oh, `<root.person>`, that is associated parent

**Reply by runningjustin** [reply]
I don't see how that's going to work. Here are my entities (with irrelevant parts removed):

```
@Entity
interface Person : BaseEntity {
    val firstName: String?
    val lastName: String?

    // user_id is nullable in database to facilitate delete, but realistically always populated
    @Key @OneToOne @OnDissociate(DissociateAction.SET_NULL) val user: User?
}
```
```
@Entity
@Table(name = "Users")
interface User : BaseEntity {
    @OneToOne @Key @Contextual @OnDissociate(DissociateAction.DELETE) val person: Person
}
```
```
@MappedSuperclass
interface BaseEntity {
    @Id @GeneratedValue(generatorType = IdGenerator::class) val id: String
    @get:JsonFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'") val createdAt: Instant
    @get:JsonFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'") val updatedAt: Instant
}
```

Do I need to insert User first, then insert Person or can I do it as proposed above even though I won't know the User's Id until after inserting person. It seems the order needs to be:
1. Insert person w/ no user_id
2. Insert user, including populating person_id from step 1 
3. Update person to set user_id from step 2

OR, since Jimmer is generating the Id using my IdGenerator:
1. Generate Ids for User and Person
2. Insert User & Person in no particular order

If there is a better way to structure my entities so I can do a single idempotent insert as above, let me know.

As always, thank you!!

**Comment by babyfish-ct** [comment]
I will release a new version later.

The is very special, it is parent association(Only upsert, never detach), not child association or middle table association

After I published the new version, parent association can also be specified by **VIOLENTLY_REPLACE**

```
save(
     Child {
         name = "c"
         parent {
             name = "p"
         }
     }
) {
    setAssociatedMode(Child::parent, AssociatedSaveMode.VIOLENTLY_REPLACE)
}
```

**Reply by runningjustin** [reply]
Haha - OK, VIOLENTLY_REPLACE it is! Let me know when the new version is out and I'll give it a try.

**Reply by babyfish-ct** [reply]
Now it only support child association and middle table association.
I forget to support parent association, please wait

**Reply by babyfish-ct** [reply]
Its java doc will be changed to

```
<p>The operation method is very simple, first clearing all old associated
     * objects<i>(If the association is NOT parent based on foreign key,
     * because parent association will NEVER be detached)</i>,
     * then reinserting new associated objects. The performance is not as good
     * as {@link #REPLACE} and may even cause event storms when using triggers.
     * However, it can accept associated objects that have neither
     * {@link org.babyfish.jimmer.sql.Id} nor {@link org.babyfish.jimmer.sql.Key}.</p>
```

Parent object will NEVER be detached!

**Comment by babyfish-ct** [comment]
Please try `0.8.173`

**Reply by runningjustin** [reply]
Trying. I must be doing something wrong here...
![image](https://github.com/user-attachments/assets/6b8dda4b-dec4-41f9-afbe-351bd7cbd109)

**Reply by babyfish-ct** [reply]
setAssociatedMode(PersonProps.PARENT, AssociatedSaveMode,VIOLENTLY_REPLACE)

**Comment by babyfish-ct** [comment]
`0.8.174`
